File: 19.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode front=head, end = head;
        for (int i=0; i <= n; ++i) {
            if (end == null) {
                return head.next;
            }
            end = end.next;
        }

        while (end != null) {
            end = end.next;
            front = front.next;
        }

        front.next = front.next.next;

        return head;
    }
}

File: 19.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        p = head
        for i in range(n):
            p = p.next

        if not p:
            return head.next

        p2 = head
        while p.next:
            p = p.next
            p2 = p2.next

        p2.next = p2.next.next

        return head
